Added description and config instruction for shared libraries. (#1187)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/about-classloading-isolation.adoc

include::_attributes.adoc[]
endif::[]

> Mule runtime engine is developed using Java and many other third party libraries. Mule apps can also contain other Java libraries, particularly modules that extend Mule to provide integration functionality, such as the HTTP connector and the XML module. These extensions can also contain other JAR libraries.

Reviewer: I'm a bit confused by the second sentence. Mule is developed using Java and other libraries, and Mule apps "can also contain other Java libraries": "can also contain" is odd because Mule does not CONTAIN the Java library but is DEVELOPED USING the library? Also, the first sentence refers to "Java libraries" but the second to "other Java libraries"? Other than what? The third sentence does the same thing: "can also contain other." And it introduces JAR instead of referring to "Java libraries"?
I'm not sure what to suggest, but here is my best guess:
```suggestion
Mule runtime engine is developed using Java and many other third-party libraries. Mule apps can also contain Java libraries, particularly modules that extend Mule to provide integration functionality, such as Anypoint Connector HTTP and the XML module. These extension modules can also contain Java (JAR) libraries.
```

Reviewer: Oops: sorry! I just open files and off I go, and I should read your very good instructions more carefully! I'll go to Line 113 now....

Additional details: All new content starts at line 113, all other changes in the file are instances where I detected that wording was odd or outdated and made changes to sync with our current standards.

}
----

The following diagram shows how resources exported by `mule-artifact.json` are applied within an app:

image::simple-classloading-diagram.png[Simple classloading diagram]

Artifact Class Loader::
    A regular Java class loader pointing to the JAR files included in the extension. This class loader loads all files and classes of the extension.

Artifact Filtering Class Loader::
    A wrapper created over the Artifact class loader, which enforces access restrictions to the extension code for foreign artifacts (the app or other plugins). It uses the content of the `mule-artifact.json` descriptor to determine what is public.

Extension Code::
    The Mule extension code. Uses the Artifact class loader (which does not have any restriction), and it is only able to locate resources of the plugin itself.

Application Code::
    The Mule app code. It uses the Artifact Filtering class loader of the extension to prevent the app from accessing restricted code or resources.

== Shared Libraries

> All the dependencies declared in the application's `pom.xml` file are visible to the application's classloader but not visible to the classloader of each connector used in the application. Because the classloading mechanism isolates each connector to prevent it from accessing classes from other connectors, you need to declare connector dependencies as shared libraries.

Reviewer 2: `All the dependencies` -> `All the 'jar' dependencies (mot 'mule-plugin's)`

> All JAR (`.jar`) dependencies declared in the application's `pom.xml` file are visible to the application's classloader but not visible to the classloader of each connector used in the application. Because the classloading mechanism isolates each connector to prevent it from accessing classes from other connectors, you need to declare connector dependencies as shared libraries.

Reviewer 3: I would change this: 

> Because the classloading mechanism isolates each connector to prevent it from accessing classes from other connectors, you need to declare connector dependencies as shared libraries.

You don't need to declare connector dependencies as shared libraries, only app's and only in some specific cases.

to something like: 
- Both the connector's and application's dependencies are loaded by different classloaders, isolating classes from each other in order to prevent access between them. However, there are cases when application dependencies need to be visible from a connector's classloader. In those cases, those dependencies need to be declared as shared libraries.

> Consider an application that uses the Java connector and the connector needs to use a class that is part of a dependency declared in the application's `pom.xml` file, this is not possible because the connector's classloader is not able to find the class. To make this class visible to the connector, you must declare the dependency that contains the class as a shared library in the Mule Maven Plugin configuration of your application's `pom.xml` file.

Reviewer 2: `jar` dependency

> Consider an application that uses the Java connector and the connector needs to use a class that is part of a JAR dependency declared in the application's `pom.xml` file, this is not possible because the connector's classloader is not able to find the class. To make this class visible to the connector, you must declare the dependency that contains the class as a shared library in the Mule Maven Plugin configuration of your application's `pom.xml` file.

Reviewer 3: 'the class' -> that class

}
----

The following diagram shows how resources exported by `mule-artifact.json` are applied within an app:

image::simple-classloading-diagram.png[Simple classloading diagram]

Artifact Class Loader::
    A regular Java class loader pointing to the JAR files included in the extension. This class loader loads all files and classes of the extension.

Artifact Filtering Class Loader::
    A wrapper created over the Artifact class loader, which enforces access restrictions to the extension code for foreign artifacts (the app or other plugins). It uses the content of the `mule-artifact.json` descriptor to determine what is public.

Extension Code::
    The Mule extension code. Uses the Artifact class loader (which does not have any restriction), and it is only able to locate resources of the plugin itself.

Application Code::
    The Mule app code. It uses the Artifact Filtering class loader of the extension to prevent the app from accessing restricted code or resources.

== Adding Dependencies to Connectors

> When the connectors in your application need access to a dependency in your project, you need to configure the dependency to be visible by the connector's classloader. There are two different ways to do this configuration:

Reviewer: ```suggestion
When the connectors in your application need access to a dependency in your project, configure the dependency to be visible by the connector's class loader. There are two different ways to do this:
```
I looked up "classloader" to add to our term list, and it looks like the general use is "class loader" (two words) and (Java) "ClassLoader." Do you agree?

> * Configure application dependencies as <<shared-libraries, Shared Libraries>>

Reviewer: ```suggestion
* Configure application dependencies as <<shared-libraries, shared libraries>>.
```

> * Configure <<additional-plugin-dependencies, Additional Plugin Dependencies>> for the connector

Reviewer: ```suggestion
* Configure <<additional-plugin-dependencies, additional plugin dependencies>> for the connector.
```

[[shared-libraries]]
=== Shared Libraries

> All dependencies (`.jar` files for example) declared in the application's `pom.xml` file are visible to the application's classloader but not visible to the classloader of each connector used in the application. The Mule runtime engine classloading mechanism isolates each connector to prevent it from accessing classes from other connectors. If a connector's classloader needs access to an application dependency, you need to declare this dependency as a shared library.

Reviewer: Our style is to use allcap names for file types but to use mono for file names that include characters plus a dot (.) extension:
```suggestion
All dependencies (JAR files, for example) declared in the application's `pom.xml` file are visible to the application's class loader but not visible to the class loader of each connector used in the application. The Mule runtime engine class-loading mechanism isolates each connector to prevent it from accessing classes from other connectors. If a connector's class loader needs access to an application dependency, declare this dependency as a shared library.
```

Reviewer: I am going to stop marking "classloader," just in case my research was faulty and until I hear from you.
If you agree with what I found, please search for and replace as appropriate.

> Consider an application that uses the Java connector and the connector needs to use a class that is part of a JAR dependency declared in the application's `pom.xml` file, this is not possible because the connector's classloader is not able to find that class. To make this class visible to the connector, you must declare the dependency that contains the class as a shared library in the Mule Maven Plugin configuration of your application's `pom.xml` file.

Reviewer: ```suggestion
Consider an application that uses Anypoint Connector for Java, and the connector needs to use a class that is part of a JAR dependency declared in the application's `pom.xml` file. However, this is not possible, because the connector's classloader is not able to find that class. To make this class visible to the connector, declare the dependency that contains the class as a shared library in the Mule Maven plugin configuration of your application's `pom.xml` file.
```

> If you use Anypoint Studio or Flow Designer to configure a connector that uses external libraries, the dependencies are automatically added as shared libraries. For example, if you add a Database Connector to your application, and then configure the connection driver using Anypoint Studio, the driver is automatically added as a shared library in your project's `pom.xml` file.

Reviewer: ```suggestion
If you use Anypoint Studio or Flow Designer to configure a connector that uses external libraries, the dependencies are automatically added as shared libraries. For example, if you add Anypoint Connector for Database to your application and then configure the connection driver using Anypoint Studio, the driver is automatically added as a shared library in your project's `pom.xml` file.
```

See xref:mmp-concept.adoc#configure-shared-libraries[Configure Shared Libraries] for configuration instructions.

> ==== Considerations

Reviewer: Does this really need to be a separate section? I'd just add it to the previous, as a Note, if necessary.

> Shared libraries are visible to all connectors in the application. Configuring shared libraries can cause issues if the shared library has different versions of the same classes that exist in other connectors or their dependencies.

Reviewer: ```suggestion
Shared libraries are visible to all connectors in the application. Configuring shared libraries can cause issues if the shared library has different versions of the same classes in other connectors or their dependencies.
```

[[additional-plugin-dependencies]]
=== Additional Plugin Dependencies

> Instead of configuring shared libraries, which makes application dependencies visible to all connectors in the application,

Reviewer: ```suggestion
Instead of configuring shared libraries to make application dependencies visible to all connectors in the application,
```

you can configure additional plugin dependencies to enrich a specific connector's classpath. This configuration enables you to effectively add new dependencies to a connector, as if the dependencies were declared in the connector's `pom.xml` file.

> Configuring additional plugin dependencies ensures that the dependencies are available only for the configured connector, and you don't need to declare them as application dependencies.

Reviewer: ```suggestion
Configuring additional plugin dependencies ensures that the dependencies affect only the configured connector, and you don't need to declare them as application dependencies.
```

> When you configure additional plugin dependencies you must specify the dependency version.

Reviewer: ```suggestion
When you configure additional plugin dependencies, you must specify the dependency version.
```

== Shared Libraries

> All dependencies (`.jar` files for example) declared in the application's `pom.xml` file are visible to the application's classloader but not visible to the classloader of each connector used in the application. The classloading mechanism isolates each connector to prevent it from accessing classes from other connectors. If a connector's classloader needs access to an application dependency, you need to declare this dependency as a shared library.

Reviewer 3: The classloading mechanism -> The Runtime classloading mechanism

If you use Anypoint Studio or Flow Designer to configure a connector that uses external libraries, the dependencies are automatically added as shared libraries. For example, if you add a Database Connector to your application, and then configure the connection driver using Anypoint Studio, the driver is automatically added as a shared library in your project's `pom.xml` file.

See xref:mmp-concept.adoc#configure-shared-libraries[Configure Shared Libraries] for configuration instructions.

=== Considerations

Shared libraries are visible to all connectors in the application. Configuring shared libraries can cause issues if the shared library has different versions of the same classes that exist in other connectors or their dependencies.

> In case you need to configure a different version of the same dependency for different connectors, configure them as plugin dependencies. Plugin dependencies are configured at connector level and are visible only to that connector.

Reviewer 3: I would change it to something like: 
For those cases, there is another option you can use to enrich a connector's classpath from an application: Additional Plugin Dependencies. 
With this approach, instead of sharing an application dependency, you can effectively add new dependencies to a connector as if they were declared in its pom.xml. 
That means, not only that those new dependencies will exclusively be available to that connector, but that the application does not have the need to declare them as it own. (Also, the dependencies version must be specified since they will no longer be available from the application dependencies declaration.)

Shared libraries are visible to all connectors in the application. Configuring shared libraries can cause issues if the shared library has different versions of the same classes that exist in other connectors or their dependencies.

== Additional Plugin Dependencies

> Instead of configuring shared libraries, which make dependencies visible to all connectors in the application,

Reviewer 3: which make dependencies -> which make application dependencies visible to al it's connectors

Shared libraries are visible to all connectors in the application. Configuring shared libraries can cause issues if the shared library has different versions of the same classes that exist in other connectors or their dependencies.

> To avoid such conflicts, you can configure additional plugin dependencies to enrich a connector's classpath. This configuration enables you to effectively add new dependencies to a connector, as if the dependencies were declared in the connector's `pom.xml` file. +

Reviewer 3: I still feel that we need to somehow state that this is a different approach to solve the same problem. You can use either <sharedLibraries> or <additionalPluginDepenndencies> but, you will rarely use both. 
That's why i would add something like: There is another configuration option....

Author: Do you think it is better to have a separate section to define additional plugin dependencies? Please check commit c3c4b1b and let me know if that's ok.

Reviewer 3: If we add a different section, i would arrange everything differently. First, showing the problem we have to solve (add dependencies to connectors) and then show the 2 solutions.

Author: I created an intro to the topic, and then presented the two solutions. 
Please check 9691c12 and let me know if the introductory section is accurate.
